File: 0207-course-schedule/0207-course-schedule.ts
function dfs(graph:number[][],isVisited:boolean[],pathVis:boolean[],index:number):boolean{
        if(pathVis[index]) return false
        isVisited[index]=true
        pathVis[index]=true
        let m=graph[index].length
        for(let i=0;i<m;i++){
            if(!isVisited[graph[index][i]]){
                if(!dfs(graph,isVisited,pathVis,graph[index][i])){
                return false
                }
            }else if(pathVis[graph[index][i]]) return false
        }
        pathVis[index]=false
        return true
    }
function createGraph(n:number,prerequisites: number[][]):number[][]{
    let graph:number[][]=new Array<number[]>(n)
    for(let i=0;i<n;i++) graph[i]=[]
    for(let i=0;i<prerequisites.length;i++){
        let indexToBeInserted=prerequisites[i][0]
        let itemToBeInserted=prerequisites[i][1]
        graph[indexToBeInserted].push(itemToBeInserted)
    }
    return graph
}

function canFinish(n: number, prerequisites: number[][]): boolean {
    let graph= createGraph(n,prerequisites)
    let isVisited:boolean[]=new Array<boolean>(n)
    let pathVis:boolean[]=new Array<boolean>(n)
    isVisited.fill(false)
    pathVis.fill(false)
    for(let i=0;i<n;i++){
        if(!isVisited[i]&&!dfs(graph,isVisited,pathVis,i)) return false
    }
    return true
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    ArrayList<ArrayList<Integer>> createGraph(int n, int[][]prerequisites){
    ArrayList<ArrayList<Integer>>graph=new ArrayList<>();
    for(int i=0;i<n;i++) graph.add(new ArrayList<Integer>());
    for(int i=0;i<prerequisites.length;i++){
        int indexToBeInserted=prerequisites[i][0];
        int itemToBeInserted=prerequisites[i][1];
        graph.get(indexToBeInserted).add(itemToBeInserted);
    }
    return graph;
    }
    public boolean canFinish(int n, int[][] prerequisites) {
        ArrayList<ArrayList<Integer>>graph=createGraph(n,prerequisites);
        Queue<Integer>q=new LinkedList();
        Stack<Integer>stack=new Stack<Integer>();
        int[]inner=new int[n];
        Arrays.fill(inner,0);
        for(int i=0;i<n;i++){
            int m=graph.get(i).size();
            for(int j=0;j<m;j++){
                inner[graph.get(i).get(j)]+=1;
            }
        }
        for(int i=0;i<n;i++){
            if(inner[i]==0){
                q.add(i);
            }
        }  
        while(!q.isEmpty()){
            int cur=q.poll();
            stack.push(cur);
            for(int i=0;i<graph.get(cur).size();i++){
                inner[graph.get(cur).get(i)]-=1;
                if(inner[graph.get(cur).get(i)]==0) q.add(graph.get(cur).get(i));
            }
        }
        return stack.size()==n;
    }
}